src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -582,19 +596,92 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+
+        return HandleRequestBodyAndCompileRequestDelegateForJsonBody(responseWritingMethodCall, factoryContext);
+    }
+
+    private static Func<object?, HttpContext, Task> HandleRequestBodyAndCompileRequestDelegateForRawBody(Expression responseWritingMethodCall, FactoryContext factoryContext)
+    {
+        Debug.Assert(factoryContext.RequestBodyParameter is not null, "factoryContext.RequestBodyParameter is null for a body parameter.");
+        Debug.Assert(factoryContext.RequestBodyParameter.Name is not null, "CreateArgument() should throw if parameter.Name is null.");
+
+        if (factoryContext.ParameterBinders.Count > 0)
+        {
+            // We need to generate the code for reading from the body before calling into the delegate
+            var continuation = Expression.Lambda<Func<object?, HttpContext, ReadOnlySequence<byte>, object?[], Task>>(
+            responseWritingMethodCall, TargetExpr, HttpContextExpr, RawBodyValueExpr, BoundValuesArrayExpr).Compile();
+
+            // Looping over arrays is faster
+            var binders = factoryContext.ParameterBinders.ToArray();
+            var count = binders.Length;
+
+            return async (target, httpContext) =>
+            {
+                // Run these first so that they can potentially read and rewind the body
+                var boundValues = new object?[count];
+
+                for (var i = 0; i < count; i++)
+                {
+                    boundValues[i] = await binders[i](httpContext);
+                }
+
+                var bodyValue = await ReadBodyAsync(httpContext);
+
+                await continuation(target, httpContext, bodyValue, boundValues);
+            };
+        }
         else
         {
-            return HandleRequestBodyAndCompileRequestDelegateForJson(responseWritingMethodCall, factoryContext);
+            // We need to generate the code for reading from the body before calling into the delegate
+            var continuation = Expression.Lambda<Func<object?, HttpContext, ReadOnlySequence<byte>, Task>>(
+            responseWritingMethodCall, TargetExpr, HttpContextExpr, RawBodyValueExpr).Compile();
+
+            return async (target, httpContext) =>
+            {
+                var bodyValue = await ReadBodyAsync(httpContext);
+
+                await continuation(target, httpContext, bodyValue);
+            };
+        }
+
+        static async ValueTask<ReadOnlySequence<byte>> ReadBodyAsync(HttpContext httpContext)
+        {
+            var feature = httpContext.Features.Get<IHttpRequestBodyDetectionFeature>();
+
+            if (feature?.CanHaveBody == true)
+            {
+                var bodyReader = httpContext.Request.BodyReader;
+
+                while (true)
+                {
+                    var result = await bodyReader.ReadAsync();
+                    var buffer = result.Buffer;
+
+                    if (result.IsCompleted)
+                    {
+                        return buffer;
```

Comment:
I think calling the route handler with the reader in an incomplete/unadvanced state is problematic. Any code accessing the `BodyReader`, either through the `HttpContext` or as a separate `PipeReader` parameter, will be unable to partially examine the body. [Any attempt will throw an InvalidOperationException](https://github.com/dotnet/runtime/blob/32a24b8b1fdf45dde7582ceb8812559e07e8454b/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L468-L471) complaining that "The examined position cannot be less than the previously examined position."

I think we're better off just fully consuming the body if you take a `ReadOnlySequence<byte>` parameter.

Reply:
I don’t want to copy but I appreciate the potential problems. I’d rather leave that with the caveats or cut the whole feature (which would make me sad but I’d get over it).

Reply:
Given this endpoint:

```C#
app.MapPost("/read-twice", async (HttpContext httpContext, ReadOnlySequence<byte> body) =>
{
    await httpContext.Request.Body.ReadAsync(new byte[1], 0, 0);
});
```

you get the following error:

```
System.InvalidOperationException: Reading is already in progress.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1ContentLengthMessageBody.ReadAsyncInternal(CancellationToken cancellationToken) in C:\dev\dotnet\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\Http1ContentLengthMessageBody.cs:line 34
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 destination, CancellationToken cancellationToken) in C:\dev\dotnet\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpRequestStream.cs:line 114
   at Program.<>c.<<<Main>$>b__0_7>d.MoveNext() in C:\dev\dotnet\aspnetcore\src\Http\samples\MinimalSample\Program.cs:line 42
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass54_1.<<HandleRequestBodyAndCompileRequestDelegateForRawBody>b__2>d.MoveNext() in C:\dev\dotnet\aspnetcore\src\Http\Http.Extensions\src\RequestDelegateFactory.cs:line 647

```

I don't think this is acceptable. We should either:

1. Consume the entire body so this read returns 0
2. Buffer the entire body so you can actually read it again
3. Poison the body so that it throws a better error than `System.InvalidOperationException: Reading is already in progress.`
4. Cut the whole feature

Reply:
OK I [pushed another commit](https://github.com/dotnet/aspnetcore/pull/39388/commits/df767a76a7794d9a9d868f3ca3659d223d4d278d) with option 1a. I set the body to Stream.Null. This fixes trying to consume the body after it's already been consumed by our binding logic.